python/ray/autoscaler/aws/config.py

```diff
@@ -380,20 +478,16 @@ def _get_key(key_name, config):
 
 
 def _client(name, config):
-    boto_config = Config(retries={"max_attempts": BOTO_MAX_RETRIES})
-    aws_credentials = config["provider"].get("aws_credentials", {})
-    return boto3.client(
-        name,
-        config["provider"]["region"],
-        config=boto_config,
-        **aws_credentials)
+    return _resource(name, config).meta.client
 
 
 def _resource(name, config):
     boto_config = Config(retries={"max_attempts": BOTO_MAX_RETRIES})
     aws_credentials = config["provider"].get("aws_credentials", {})
-    return boto3.resource(
+    return _RESOURCE_CACHE.setdefault(
         name,
-        config["provider"]["region"],
-        config=boto_config,
-        **aws_credentials)
+        boto3.resource(
```

Comment:
I just realized, doesn't this have the same issue as `_get_vpc_id_or_die` with the default value not being lazy?

Sorry, I think I asked you to change to what is here now in an earlier review >.<

Reply:
Yep - same issue. It calls boto3 to get the resource, then finds out that it already exists in the cache and discards it! 😝

Updated to:
``` python
    if name not in _RESOURCE_CACHE:
        _RESOURCE_CACHE[name] = boto3.resource(
            name,
            config["provider"]["region"],
            config=boto_config,
            **aws_credentials,
        )
    return _RESOURCE_CACHE[name]
```